---
title: "ex_3_Yuval_Ben_Eliahu"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r 1}
library("compGenomRData")
counts_file <- system.file("extdata/rna-seq/SRP029880.raw_counts.tsv",
                           package = "compGenomRData")
coldata_file <- system.file("extdata/rna-seq/SRP029880.colData.tsv",
                            package = "compGenomRData")

counts <- read.table(counts_file, header = T, sep = '\t')
#remove the 'width' column
countData <- as.matrix(subset(counts, select = c(-width)))

colData <- read.table(coldata_file, header = T, sep = '\t',
                      stringsAsFactors = TRUE) 

# Set up a DESeqDataSet object.
library(DESeq2)

#create a DESeq dataset object from the count matrix and the colData
dds <- DESeqDataSetFromMatrix(countData = countData,
                              colData = colData,
                              design = ~ group)

#Filter out genes with low counts.
filteration <- rowSums(DESeq2::counts(dds)) >= 10

dds <- dds[filteration, ]

dds <- DESeq(dds)
```

2.
```{r 2}
DESeq2::plotDispEsts(dds)
#the dispersion plot show the relationship between the variance in gene expression and the mean value.
```

3.
```{r 3}
#lfcThreshold default value is 0. changing it to 1 changes the log2 fold change threshold to 1. It will also incorporate the results under Walt tests.

DEresults=results(dds,contrast = c("group", 'CASE', 'CTRL'))
DEresults <- DEresults[order(DEresults$pvalue),]
DEresults2=results(dds,contrast = c("group", 'CASE', 'CTRL'), lfcThreshold=1)
DEresults2 <- DEresults2[order(DEresults$pvalue),]
```

4.
independent filtering is used to filter out genes with a low read counts from the dataset to reduce false positives. It keeps genes with average CPM (count per million) higher than the minimum CPM.
if we don't use it then our results will have more False positives and will be less significant.

5.
```{r 4}
library("edgeR")

GenewiseCounts <- countData
dgList <- DGEList(GenewiseCounts[,-1],genes=GenewiseCounts[,1,drop=FALSE])

countsPerMillion <- cpm(dgList)
countCheck <- countsPerMillion > 1
keep <- which(rowSums(countCheck) >= 2)
dgList <- dgList[keep,]
dgList <- calcNormFactors(dgList, method="TMM")

sampleType<- rep("N", ncol(dgList)) 
sampleType[grep("T", colnames(dgList))] <- "T"
sampleReplicate <- paste("S", rep(1:3, each=2), sep="")
```

6.
```{r 5}
library("compcodeR")
count.matrix <- countData
sample.annot <- data.frame(colData)
names(sample.annot)[1] <- "condition"

info.parameters <- list(dataset = "mytestdata", uID = "123456")
cpd <- compData(count.matrix = count.matrix, 
                sample.annotations = sample.annot, 
                info.parameters = info.parameters)

saveRDS(cpd, file = "cpd.rds")

runDiffExp(data.file = "cpd.rds", 
           result.extent = "voom.limma", Rmdfunction = "voom.limma.createRmd", 
           output.directory = ".", norm.method = "TMM")
runDiffExp(data.file = "cpd.rds", 
           result.extent = "edgeR.exact", Rmdfunction = "edgeR.exact.createRmd", 
           output.directory = ".", norm.method = "TMM", 
           trend.method = "movingave", disp.type = "tagwise")
runDiffExp(data.file = "cpd.rds", result.extent = "ttest", 
           Rmdfunction = "ttest.createRmd", 
           output.directory = ".", norm.method = "TMM")
```

```{r 6}
runComparisonGUI(input.directories = ".", 
                 output.directory = ".", recursive = FALSE)
```