---
title: "ex_4_Yuval_Ben_Eliahu"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

### Functional enrichment analysis

Loading the required packages:
```{r message = F, warning=FALSE}
library(compGenomRData)
library(DESeq2)
library(gProfileR)
library(gage)
```

Loading the data:
```{r message = F, warning=FALSE}
counts_file <- system.file("extdata/rna-seq/SRP029880.raw_counts.tsv",
                           package = "compGenomRData")
coldata_file <- system.file("extdata/rna-seq/SRP029880.colData.tsv",
                            package = "compGenomRData")
counts <- as.matrix(read.table(counts_file, header = T, sep = '\t'))

#remove the 'width' column
countData <- as.matrix(subset(counts, select = c(-width)))

#define the experimental setup
colData <- read.table(coldata_file, header = T, sep = '\t',
                      stringsAsFactors = TRUE)

#define the design formula
designFormula <- "~ group"

#create a DESeq dataset object from the count matrix and the colData
dds <- DESeqDataSetFromMatrix(countData = countData,
                              colData = colData,
                              design = as.formula(designFormula))
dds <- DESeq(dds)

DEresults = results(dds, contrast = c("group", 'CASE', 'CTRL'))
```

Let’s select the genes that are significantly differentially expressed between the case and control samples. Let’s extract genes that have an adjusted p-value below 0.1 and that show a 2-fold change (either negative or positive) in the case compared to control. We will then feed this gene set into the gProfileR function.

```{r message = F, warning=FALSE}
#remove genes with NA values
DE <- DEresults[!is.na(DEresults$padj),]

#select genes with adjusted p-values below 0.1
DE <- DE[DE$padj < 0.1,]

#select genes with absolute log2 fold change above 1 (two-fold change)
DE <- DE[abs(DE$log2FoldChange) > 1,]

#get the list of genes of interest
genesOfInterest <- rownames(DE)

#calculate enriched GO terms
goResults <- gprofiler(query = genesOfInterest,
                       organism = 'hsapiens',
                       src_filter = 'GO',
                       hier_filtering = 'moderate')
```

1. Re-run gProfileR, this time using pathway annotations such as KEGG, REACTOME, and protein complex databases such as CORUM, in addition to the GO terms. Sort the resulting tables by columns `precision` and/or `recall`. How do the top GO terms change when sorted for `precision`, `recall`, or `p.value`? hint: use `order()` for sorting. [Difficulty: **Beginner**]

```{r}
KEGGResults <- gprofiler(query = genesOfInterest,
                       organism = 'hsapiens',
                       src_filter = 'KEGG',
                       hier_filtering = 'moderate')
REACResults <- gprofiler(query = genesOfInterest,
                       organism = 'hsapiens',
                       src_filter = 'REAC',
                       hier_filtering = 'moderate')
CORUMResults <- gprofiler(query = genesOfInterest,
                       organism = 'hsapiens',
                       src_filter = 'CORUM',
                       hier_filtering = 'moderate')
```

```{r}
KEGGResults.Prec=KEGGResults[order(KEGGResults$precision),]  
KEGGResults.rec=KEGGResults[order(KEGGResults$recall),]  

REACResults.Prec=REACResults[order(REACResults$precision),]  
REACResults.rec=REACResults[order(REACResults$recall),] 

CORUMResults.Prec=CORUMResults[order(CORUMResults$precision),]  
CORUMResults.rec=CORUMResults[order(CORUMResults$recall),]  

goResults.Prec=goResults[order(goResults$precision),]  
goResults.rec=goResults[order(goResults$recall),]  
goResults.pval=goResults[order(goResults$p.value),]  
```

#### Gene set enrichment analysis

We use the bioconductor package gage to demonstrate how to do GSEA using normalized expression data of the samples as input.

```{r}
#Let's define the first gene set as the list of genes from one of the
#significant GO terms found in the GO analysis. order go results by pvalue
goResults <- goResults[order(goResults$p.value),]

#restrict the terms that have at most 100 genes overlapping with the query
go <- goResults[goResults$overlap.size < 100,]

# use the top term from this table to create a gene set
geneSet1 <- unlist(strsplit(go[1,]$intersection, ','))

#Define another gene set by just randomly selecting 25 genes from the counts
#table get normalized counts from DESeq2 results
normalizedCounts <- DESeq2::counts(dds, normalized = TRUE)

geneSet2 <- sample(rownames(normalizedCounts), 25)

geneSets <- list('top_GO_term' = geneSet1,
                 'random_set' = geneSet2)

# Using the defined gene sets, we’d like to do a group comparison between the case
# samples with respect to the control samples.

#Use the normalized counts to carry out a GSEA.
gseaResults <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets, compare = 'as.group')
```

2. Repeat the gene set enrichment analysis by trying different options for the `compare` argument of the `GAGE:gage`function. How do the results differ? [Difficulty: **Beginner**]

The gene set show a much more significant up-regulation when using the paired, unpaired and 1ongroup options when compared to the as.group option (much lower p.value).

```{r}
gseaResults.paired <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets, compare = 'paired')

gseaResults.unpaired <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets, compare = 'unpaired')

gseaResults.1ongroup <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets, compare = '1ongroup')

gseaResults$greater
gseaResults.paired$greater
gseaResults.unpaired$greater
gseaResults.1ongroup$greater
```

3. Make a scatter plot of GO term sizes and obtained p-values by setting the `gProfiler::gprofiler` argument `significant = FALSE`. Is there a correlation of term sizes and p-values? (Hint: Take -log10 of p-values). If so, how can this bias be mitigated? [Difficulty: **Intermediate**]

```{r}
goResults.false <- gprofiler(query = genesOfInterest,
                       organism = 'hsapiens',
                       src_filter = 'GO',
                       hier_filtering = 'moderate',significant=F)

goResults.false[,c(3)] <- -log10(goResults.false[,c(3)])
```

```{r}

library(ggplot2)

ggplot(goResults.false, aes(x=term.size, y=p.value)) +
geom_point()
```


4. Do a gene-set enrichment analysis using gene sets from top 10 GO terms. [Difficulty: **Intermediate**]

```{r}
goResults.top10 <- goResults[goResults$overlap.size < 100,]
goResults.top10 <- goResults.top10[1:10,]

geneSet.top10 <- unlist(strsplit(goResults.top10$intersection, ','))

normalizedCounts <- DESeq2::counts(dds, normalized = TRUE)

geneSet.random <- sample(rownames(normalizedCounts), 600)

geneSet <- list('top_10' = geneSet.top10,
                 'random_set' = geneSet.random)

gseaResult <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSet, compare = 'as.group')
```

5. What are the other available R packages that can carry out gene set enrichment analysis for RNA-seq datasets? [Difficulty: **Intermediate**]

```{r}
"We can use fgsea, goseq, clusterProfiler, topGO 
```

6.  Use the topGO package (https://bioconductor.org/packages/release/bioc/html/topGO.html) to re-do the GO term analysis. Compare and contrast the results with what has been obtained using the `gProfileR` package. Which tool is faster, `gProfileR` or topGO? Why? [Difficulty: **Advanced**]

```{r}
library(topGO)
library(ALL)
data(ALL)
data(geneList)
affyLib <- paste(annotation(ALL), "db", sep = ".")
library(package = affyLib, character.only = TRUE)
sum(topDiffGenes(geneList))
sampleGOdata <- new("topGOdata",
description = "Simple session", ontology = "BP",
allGenes = geneList, geneSel = topDiffGenes,
nodeSize = 10, annot = annFUN.db, affyLib = affyLib)
```

```{r}
sampleGOdata
```

```{r}
resultFisher <- runTest(sampleGOdata, algorithm = "classic", statistic = "fisher")
resultFisher
```

7. Given a gene set annotated for human, how can it be utilized to work on _C. elegans_ data? (Hint: See `biomaRt::getLDS`). [Difficulty: **Advanced**]

We can use the getLDS() function to link 2 datasets and construct a query over them. This way we can retrive homology data from the human gene set across to the C.elegans specie. For example:

```{r}
library(biomaRt)
human = useMart("ensembl", dataset = "hsapiens_gene_ensembl")
C.elegans = useMart("ensembl", dataset = "celegans_gene_ensembl") 
getLDS(attributes = c("hgnc_symbol","chromosome_name", "start_position"), 
    filters = "hgnc_symbol", values = "TP53", mart = human, 
    attributesL = c("chromosome_name","start_position"), martL = C.elegans)
```

8. Import curated pathway gene sets with Entrez identifiers from the [MSIGDB database](http://software.broadinstitute.org/gsea/msigdb/collections.jsp) and re-do the GSEA for all curated gene sets. [Difficulty: **Advanced**]

```{r}
#I downloaded the gene set "BIOCARTA_41BB_PATHWAY" in text format
library("readxl")
geneset.Bio=read_excel("C:\\Users\\besterlab\\OneDrive\\Yuval\\OneDrive - Technion\\Master\\Courses\\Computational Genomics\\ex4\\geneset.xlsx")

geneset.Bio <- geneset.Bio[2:19,]

geneset.Bio<-as.data.frame(geneset.Bio)

geneset.Bio<-as.character(geneset.Bio[,1])

geneSets.Bio <- list('top_GO_term' = geneset.Bio,
                 'random_set' = geneSet2)

gseaResults.Bio <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets.Bio, compare = 'as.group')

```